components/sup/src/manager/mod.rs

```diff
@@ -523,9 +528,11 @@ impl Manager {
             return;
         }
 
-        self.gossip_latest_service_rumor(&service);
-        if service.topology == Topology::Leader {
-            self.butterfly.start_election(&service.service_group, 0);
+        if let Some(ref butterfly) = self.butterfly {
+            self.gossip_latest_service_rumor(&service);
```

Comment:
There's an `if let Some(…) = self.butterfly` conditional both here *and* inside `gossip_latest_service_rumor`. For simplicity and testability, I'm a fan of having code be either:

1. Logic with loops and branches that encode behavior
1. Lists of mostly unconditional commands that plumb together different chunks of (1)

Try to minimize mixing (1) and (2). The (1)-type code is where the testing should be focused, so making that code as purely functional as possible, with a minimum of side effects or superfluous state carried around will make the testing easiest. The (2)-type code won't really have any behavior to test and with a language like rust, errors in plumbing will nearly all be caught by the compiler.

Despite it having a conditional in the form of `if let`, I would call this more plumbing (2) than behavioral logic (1), so keeping it at the higher level and making `gossip_latest_service_rumor` more pure by passing it the specific data it needs to operate on rather than all of `self` should eliminate the redundancy.

Reply:
Here's the general pattern I think we should try to avoid:
```rust
        if let Some(ref butterfly) = self.butterfly {
            self.some_function_that_needs_butterfly(…);
```
This is how we end up exploding the number of redundant `if let` conditionals with `self.butterfly`. Inside any of those conditionals, if we're calling functions that need butterfly, we should provide it as a parameter. This gives is a purer, more testable function and avoids the need for redundant checks.